src/main/java/net/fabricmc/loom/configuration/providers/mappings/TinyMappingsService.java

```diff
@@ -43,13 +48,37 @@ public TinyMappingsService(Path tinyMappings) {
 		} catch (IOException e) {
 			throw new UncheckedIOException("Failed to read mappings", e);
 		}
+
+		this.defaultSrcNs = this.mappingTree.getSrcNamespace();
+		this.lastSrcNs = this.defaultSrcNs;
 	}
 
 	public static synchronized TinyMappingsService create(SharedServiceManager serviceManager, Path tinyMappings) {
 		return serviceManager.getOrCreateService("TinyMappingsService:" + tinyMappings.toAbsolutePath(), () -> new TinyMappingsService(tinyMappings));
 	}
 
 	public MemoryMappingTree getMappingTree() {
+		return getMappingTree(defaultSrcNs);
+	}
+
+	public MemoryMappingTree getMappingTree(MappingsNamespace srcNs) {
+		return getMappingTree(srcNs.toString());
+	}
+
+	public MemoryMappingTree getMappingTree(String srcNs) {
+		if (!srcNs.equals(lastSrcNs)) {
+			MemoryMappingTree tree = new MemoryMappingTree();
+
+			try {
+				mappingTree.accept(new MappingSourceNsSwitch(tree, srcNs));
+			} catch (IOException e) {
+				throw new UncheckedIOException("Failed to switch source namespace", e);
+			}
+
+			mappingTree = tree;
+			lastSrcNs = srcNs;
+		}
```

Comment:
This feels wrong, I think you should have seperate instances of this service for each namespace. `getOrCreateService` takes a key, that should likely include the namespace.